### PR TITLE
Fix DoS #183

### PR DESCRIPTION
**Pull request by junorouse**
Print warning message when language is a NULL pointer.

### CODE REVIEW

File: code/commands/view-commands.cpp [hunk 1]
```diff
@@ -504,7 +504,11 @@ void Gobby::ViewCommands::on_menu_language_changed()
 
 		// The language should exist by construction, if the languages
 		// available in the language manager don't change at runtime
-		g_assert(language != NULL);
+		if(language != NULL) {
```
Comment (aburgm) [inline code/commands/view-commands.cpp:507]: This should be `language == NULL` or you are doing the exact opposite of your intention :)

**Review by aburgm — CHANGES_REQUESTED** [review]
If you are concerned about this, you might also want to consider handling the case of language_id being empty, which would probably lead to the same segmentation fault?

File: code/commands/view-commands.cpp [hunk 2]
```diff
@@ -504,7 +504,11 @@ void Gobby::ViewCommands::on_menu_language_changed()
...
+			g_warning("Language \"%s\" is not known to GtkSourceView",
+				language_id.c_str());
+			return;
```
Comment (iamahuman) [inline code/commands/view-commands.cpp:510]: Are you sure you want to return from the function here? `NULL` is a perfectly legal value for `language` in this case, as can be seen by the initial value of `language` which will be used as-is when the `language_id.empty()` condition above is false.

Perhaps simply taking out `g_assert(language != NULL);` would suffice, and the user is not going to look at the GLib warning logs anyway.

**Comment by aburgm** [comment]
I'm sorry, I just had another look at #183, and I don't believe this is a correct fix for the issue. Looking at the location of the crash (from #183), we got:

```
Thread 1 "gobby-0.5" received signal SIGSEGV, Segmentation fault.
0x000055555561d681 in Gobby::TextSessionView::set_language (this=0x0, language=language@entry=0x0) at code/core/textsessionview.cpp:429
429		gtk_source_buffer_set_language(m_buffer, language);
```

I don't think the problem is that language is NULL, but that `this` is NULL. In other words, when you make your dbus call with an existing language ID, I would still expect the application to crash, and it is the `g_assert(m_current_view != NULL)` that needs to be protected with a proper guard, not the language. A NULL language is actually valid, it just disables syntax highlighting altogether (see [docs](https://developer.gnome.org/gtksourceview/stable/GtkSourceBuffer.html#gtk-source-buffer-set-language)).

I'm not sure why `m_current_view` would be NULL; are you maybe making the call when there is no document open in Gobby? I think the menu item is greyed out in this case, but maybe the d-bus call can still be made. In that case I would expect many other "invalid" menu actions could be triggered that way as well, leading to similar problems.

**Comment by junorouse** [comment]
@aburgm Sorry, I mislead the root cause of this crash.

> are you maybe making the call when there is no document open in Gobby?
- Yes.

> In that case I would expect many other "invalid" menu actions could be triggered that way as well, leading to similar problems.

So the proper solutions might be ...

1.  Patch all the `g_assert` codes to print warning message instead of crashing.

```
juno@ubuntu:~/Desktop/gobby$ grep -R 'g_assert(m_current_view != NULL);' .
./code/commands/view-commands.cpp:	g_assert(m_current_view != NULL);
./code/commands/edit-commands.cpp:	g_assert(m_current_view != NULL);
./code/commands/edit-commands.cpp:	g_assert(m_current_view != NULL);
./code/commands/edit-commands.cpp:	g_assert(m_current_view != NULL);
./code/commands/edit-commands.cpp:	g_assert(m_current_view != NULL);
./code/commands/edit-commands.cpp:	g_assert(m_current_view != NULL);
./code/commands/edit-commands.cpp:	g_assert(m_current_view != NULL);
./code/commands/edit-commands.cpp:	g_assert(m_current_view != NULL);
./code/commands/edit-commands.cpp:	g_assert(m_current_view != NULL);
./code/commands/edit-commands.cpp:	g_assert(m_current_view != NULL);
./code/commands/edit-commands.cpp:	g_assert(m_current_view != NULL);
./code/dialogs/goto-dialog.cpp:		g_assert(m_current_view != NULL);
./code/dialogs/goto-dialog.cpp:	g_assert(m_current_view != NULL);
```

2. to do nothing and maintain the current logic. (won't fix)

**Comment by aburgm** [comment]
> Patch all the g_assert codes to print warning message instead of crashing.

This would be fine with me.

> to do nothing and maintain the current logic. (won't fix)

But this too. :)

**Comment by junorouse** [comment]
@aburgm I made a patch with the former one.

**Comment by junorouse** [comment]
@aburgm ping!

**Comment by aburgm** [comment]
Looks good to me, sorry for the delay!

**Comment by carnil** [comment]
Looks this got [CVE-2020-35450](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-35450) assigned.

**Comment by abergmann** [comment]
[CVE-2020-35450](https://nvd.nist.gov/vuln/detail/CVE-2020-35450) was assigned to this issue.

**Pull request merged**